uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js

```diff
@@ -0,0 +1,283 @@
+    };
+
+    /**
+     * Converts given date to moment object with server format
+     *
+     * @param {Date/String} date from server
+     * @return {Date} Moment date
+     */
+    this._getDateObjectWithFormat = function (date) {
+      return moment(date, sharedSettings.serverDateFormat).clone();
+    };
+
+    /**
+     * Find the month which matches with the sent date
+     * and return the related object
+     *
+     * @param {object} date
+     * @param {array} months
+     * @return {object}
+     */
+    this._getMonthObjectByDate = function (date, months) {
+      return _.find(months, function (month) {
+        return (month.month === date.month()) && (month.year === date.year());
+      });
+    };
+
+    /**
+     * Returns the styles for a specific leaveRequest
+     * which will be used in the view for each date
+     *
+     * @param  {object} leaveRequest
+     * @param  {object} dateObj - Date UI object which handles look of a calendar cell
+     * @return {object}
+     */
+    this._getStyles = function (leaveRequest, dateObj) {
+      var absenceType,
+        status = this._leaveRequestStatuses[leaveRequest.status_id];
+
+      if (status.name === 'waiting_approval'
+        || status.name === 'approved'
+        || status.name === 'admin_approved') {
+        absenceType = _.find(this.absenceTypes, function (absenceType) {
+          return absenceType.id == leaveRequest.type_id;
+        });
+
+        //If Balance change is positive, mark as Accrued TOIL
+        if (leaveRequest.balance_change > 0) {
+          dateObj.UI.isAccruedTOIL = true;
+          return {
+            border: '1px solid ' + absenceType.color
+          };
+        }
+
+        return {
+          backgroundColor: absenceType.color,
+          borderColor: absenceType.color
+        };
+      }
+    };
+
+    /**
+     * Returns whether a date is of a specific type
+     * half_day_am or half_day_pm
+     *
+     * @param  {string} name
+     * @param  {object} leaveRequest
+     * @param  {string} date
+     *
+     * @return {boolean}
+     */
+    this._isDayType = function (name, leaveRequest, date) {
+      var dayType = this._dayTypes[name];
+
+      if (moment(date).isSame(leaveRequest.from_date)) {
+        return dayType.value == leaveRequest.from_date_type;
+      }
+
+      if (moment(date).isSame(leaveRequest.to_date)) {
+        return dayType.value == leaveRequest.to_date_type;
+      }
+    };
+
+    /**
+     * Returns whether a leaveRequest is pending approval
+     *
+     * @param  {object} leaveRequest
+     * @return {boolean}
+     */
+    this._isPendingApproval = function (leaveRequest) {
+      var status = this._leaveRequestStatuses[leaveRequest.status_id];
+
+      return status.name === 'waiting_approval';
+    };
+
+    /**
+     * Loads the absence periods
+     *
+     * @return {Promise}
+     */
+    this._loadAbsencePeriods = function () {
+      var self = this;
```

Comment:
Do not using `self` when possible, we have `bind` to avoid it

```js
this._loadAbsencePeriods = function () {
  return AbsencePeriod.all()
    .then(function (absencePeriods) {
      this.absencePeriods = absencePeriods;
      this.selectedPeriod = _.find(this.absencePeriods, function (period) {
        return !!period.current;
      });

      this._fetchMonthsFromPeriod();
    }.bind(this));
};
```